src/java/htsjdk/samtools/sra/SRALazyRecord.java

```diff
@@ -0,0 +1,941 @@
+            this.getterImplName = getterImplName;
+        }
+    }
+
+    private static Map<LazyFlag, FlagGetter> lazyFlagsGetters;
+    static
+    {
+        lazyFlagsGetters = new HashMap<LazyFlag, FlagGetter>();
+        lazyFlagsGetters.put(LazyFlag.READ_NEGATIVE_STRAND, new FlagGetter("getReadNegativeStrandFlag", true));
+        lazyFlagsGetters.put(LazyFlag.READ_PAIRED, new FlagGetter("getReadPairedFlag", true));
+        lazyFlagsGetters.put(LazyFlag.NOT_PRIMARY_ALIGNMENT, new FlagGetter("getNotPrimaryAlignmentFlag", true));
+        lazyFlagsGetters.put(LazyFlag.MATE_NEGATIVE_STRAND, new FlagGetter("getMateNegativeStrandFlag", false));
+        lazyFlagsGetters.put(LazyFlag.MATE_UNMAPPED, new FlagGetter("getMateUnmappedFlag", false));
+        lazyFlagsGetters.put(LazyFlag.FIRST_OF_PAIR, new FlagGetter("getFirstOfPairFlag", false));
+        lazyFlagsGetters.put(LazyFlag.SECOND_OF_PAIR, new FlagGetter("getSecondOfPairFlag", false));
+    }
+
+    private static Map<Short, AttributeTagInfo> lazyAttributeTags;
+    static
+    {
+        lazyAttributeTags = new HashMap<Short, AttributeTagInfo>();
+        lazyAttributeTags.put(SAMTagUtil.getSingleton().RG, new AttributeTagInfo(LazyAttribute.RG, "getAttributeGroupNameImpl"));
+    }
+
+    public SRALazyRecord(final SAMFileHeader header, boolean isAligned, ReadCollection run, Object sraIterator, String readId, String alignmentId) {
+        super(header);
+
+        if (!isAligned) {
+            throw new RuntimeException("Wrong constructor called. Call constructor which accepts unalignedReadFragmentIndex");
+        }
+        if (!(sraIterator instanceof AlignmentIterator)) {
+            throw new RuntimeException("Wrong SRA iterator provided. Expected AlignmentIterator");
+        }
+
+        this.isAligned = true;
+        this.run = run;
+        this.sraIterator = sraIterator;
+        this.sraReadId = readId;
+        this.sraAlignmentId = alignmentId;
+        setReadName(readId);
+        setReadUnmappedFlag(false);
+    }
+
+    public SRALazyRecord(final SAMFileHeader header, boolean isAligned, ReadCollection run, Object sraIterator, String readId, int unalignedReadFragmentIndex) {
+        super(header);
+
+        if (isAligned) {
+            throw new RuntimeException("Wrong constructor called. Call constructor which accepts alignmentId");
+        }
+        if (!(sraIterator instanceof ReadIterator)) {
+            throw new RuntimeException("Wrong SRA iterator provided. Expected ReadIterator");
+        }
+
+        this.isAligned = false;
+        this.run = run;
+        this.sraIterator = sraIterator;
+        this.sraReadId = readId;
+        this.unalignedReadFragmentIndex = unalignedReadFragmentIndex;
+        setReadName(readId);
+        setReadUnmappedFlag(true);
+    }
+
+    public void detachFromIterator() {
+        sraIterator = null;
+    }
+
+    // ===== fields =====
+
+    @Override
+    public int getAlignmentStart() {
```

Comment:
These methods are not adequately tested via the existing tests in `SRATest`. Eg., you are not even testing the `set*` methods at all, there are untested `get*` methods like `getReferenceIndex()`, and the methods that are tested are not tested with a sufficient range of possible inputs to guarantee correctness in all cases. To ensure that each method is fully tested, it's best to write a standalone unit test for each `get/set` pair of methods.